src/interpreter/mod.rs

```diff
@@ -389,12 +389,16 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
             StackPopCleanup::Goto(target) => self.goto_block(target),
             StackPopCleanup::None => {},
         }
-        // check that all locals have been deallocated through StorageDead
+        // deallocate all locals that are backed by an allocation
         for (i, local) in frame.locals.into_iter().enumerate() {
             if let Some(Value::ByRef(ptr)) = local {
                 trace!("deallocating local {}: {:?}", i + 1, ptr);
                 self.memory.dump(ptr.alloc_id);
                 match self.memory.deallocate(ptr) {
+                    // Any frozen memory means that it belongs to a constant or something referenced
+                    // by a constant. We could alternatively check whether the alloc_id is frozen
+                    // before calling deallocate, but this is much simpler and is probably the
+                    // rare case.
```

Comment:
No worries. :P

The thing above that freezes all reachable memory is actually freezing all memory reachable from a global as part of `StackPopCleanup::Freeze`... this is the part I wasn't considering.

I was thinking of interpreting a regular function. In that case, I'm still certain the locals will never be frozen. So the `ByRef` are definitely not *always* frozen.

But when we're interpreting a global initializer function, then we freeze everything reachable from the return value (the global) and only free the locals that were true temporaries. For example, given `const X: &'static usize = &(1 + 2); X` if the `1` and `2` introduced `ByRef` locals (not true, but bear with me) then we would free those locals but not the local storing the `ByRef` with `3` in it.

Your comment even explained exactly what's happening, but I totally blanked on the concept that globals *have their own locals* when they're being initialized.